File: src/07_Dynamic_Programming/dp_knapsack.cpp
#pragma GCC optimize ("03")
#include <bits/stdc++.h>
#include "../compete.hpp"
using namespace std;


int w[] = {1, 3, 3, 5};
const int n = 4;
const int max_w = 12;
bool poss_x_k[max_w + 1][n + 1];


// O(nx) time, n: number of values, x: sum
// O(nx) space
void solve_1()
{
  Timer t(__func__);

  memset(poss_x_k, false, sizeof(poss_x_k));
  // 0 total weight is possible with 0 weights
  poss_x_k[0][0] = true;

  for (int k = 1; k <= n; ++k) {
    for (int x = 0; x <= max_w; ++x) {
      // w index must be offset by -1, because the 0th index of poss corre-
      // sponds to no weights. Namely, the 1st index of poss corresponds to
      // the 0th weight at w.
      if (x - w[k - 1] >= 0)
      poss_x_k[x][k] |= poss_x_k[x - w[k - 1]][k - 1];
      poss_x_k[x][k] |= poss_x_k[x][k - 1];
    }
  }
}


void print_solve_1()
{
  cout << "  | ";
  for (int x = 0; x <= max_w; ++x) {
    cout << right << setw(2) << x << ' ';
  }
  cout << '\n';
  cout << "0 | ";
  for (int x = 0; x <= max_w; ++x) {
    cout << right << setw(2) << (poss_x_k[x][0] ? "X": " ") << ' ';
  }
  cout << '\n';

  for (int k = 1; k <= n; ++k) {
    cout << w[k - 1] << " | ";
    for (int x = 0; x <= max_w; ++x) {
      cout << right << setw(2) << (poss_x_k[x][k] ? "X": " ") << ' ';
    }
    cout << '\n';
  }
}


bool poss_x[max_w + 1];


// O(nx) time, n: number of values, x: sum
// O(x)  space
void solve_2()
{
  Timer t(__func__);

  memset(poss_x, false, sizeof(poss_x));
  // 0 total weight is possible with 0 weights
  poss_x[0] = true;

  for (int k = 0; k < n; ++k)
    for (int x = max_w; x >= 0; --x)
      if (poss_x[x])
        poss_x[x + w[k]] = true;
}

void print_solve_2()
{
  cout << "  | ";
  for (int x = 0; x <= max_w; ++x) {
    cout << right << setw(2) << x << ' ';
  }
  cout << '\n';

  cout << "  | ";
  for (int x = 0; x <= max_w; ++x) {
    cout << right << setw(2) << (poss_x[x] ? "X": " ") << ' ';
  }
  cout << '\n';
}


int main()
{
  solve_1();
  print_solve_1();
  cout << '\n';
  solve_2();
  print_solve_2();
}

File: src/07_Dynamic_Programming/dp_knapsack.py
weights = [1, 3, 3, 5]
n = len(weights)
max_w = sum(weights)


def print_possible(possible):
  print(' '.join(map(str, range(max_w + 1))))
  fmt = "{0:>{1}}"
  print(' '.join(fmt.format('X' if possible[x] else ' ', len(str(x)))
                 for x in range(max_w + 1)))


# O(nx) time, n: number of weights, x: max-sum
# O(nx) space
def solve_2d():
  # possible[x][k] is True if x can be constructed with the first k weights
  possible = [[False for k in range(n + 1)]
              for x in range(max_w + 1)]
  # 0 total weight is possible with 0 weights
  possible[0][0] = True

  for k in range(1, n + 1):
    for x in range(max_w + 1):
      if x - weights[k - 1] >= 0:
        # eigher weight k is used
        # (index is k in possible, k-1 in weights)
        possible[x][k] |= possible[x - weights[k - 1]][k - 1]
      # or not
      possible[x][k] |= possible[x][k - 1]

  return [p[n] for p in possible]


# O(nx) time, n: number of weights, x: max-sum
# O(x) space
def solve_1d():
  # possible[x] is True if x can be constructed
  possible = [False] * (max_w + 1)
  # 0 total weight is possible with 0 weights
  possible[0] = True

  for k in range(n):
    for x in range(max_w, -1, -1):
      if possible[x]:
        possible[x + weights[k]] = True

  return possible


if __name__ == '__main__':
  # 0 1 2 3 4 5 6 7 8 9 10 11 12
  # x x   x x x x x x x     x  x
  print_possible(solve_2d())
  print()
  print_possible(solve_1d())
